commands_test.go

```diff
@@ -6369,6 +6378,118 @@ var _ = Describe("Commands", func() {
+				lib1.Functions[0].Description, lib1.Functions[0].Flags[0], lib1.Functions[0].Flags[1])
+
+			err := client.FunctionLoad(ctx, lib1Code).Err()
+			Expect(err).NotTo(HaveOccurred())
+
+			// This function doesn't have a "no-writes" flag
+			x := client.FCallRo(ctx, lib1.Functions[0].Name, []string{"my_hash"}, "a", 1, "b", 2)
+
+			Expect(x.Err()).To(HaveOccurred())
+
+			lib2Code = fmt.Sprintf(lib2.Code, lib2.Name, lib2.Functions[0].Name, lib2.Functions[1].Name,
+				lib2.Functions[1].Description, lib2.Functions[1].Flags[0])
+
+			// This function has a "no-writes" flag
+			err = client.FunctionLoad(ctx, lib2Code).Err()
+			Expect(err).NotTo(HaveOccurred())
+
+			x = client.FCallRo(ctx, lib2.Functions[1].Name, []string{})
+
+			Expect(x.Err()).NotTo(HaveOccurred())
+			Expect(x.Text()).To(Equal("Function 2"))
+		})
+
+		It("Shows function stats", func() {
+			// The 5M iterations in the f1 function should block for around 3 seconds when executed,
+			// depending on how busy the processor is
+			lib := redis.Library{
+				Name:   "mylib1",
+				Engine: "LUA",
+				Functions: []redis.Function{
+					{
+						Name:        "lib1_func1",
+						Description: "This is the func-1 of lib 1",
+						Flags:       []string{"allow-stale"},
+					},
+				},
+				Code: `#!lua name=%s
+					
+					local function f1(keys, args)
+                     local t1 = redis.call('TIME')[1]
+                     for i = 5000000, 1,-1 do 
+                     redis.call('SET', 'a', i)
+                     end
+                     local t2 = redis.call('TIME')[1]
+               
+                     redis.log(redis.LOG_NOTICE, t2-t1)
+                     return t2 - t1
+					end
+
+					redis.register_function{
+						function_name='%s',
+						description ='%s',
+						callback=f1,
+						flags={'%s'}
+					}`,
+			}
+			libCode := fmt.Sprintf(lib.Code, lib.Name, lib.Functions[0].Name,
+				lib.Functions[0].Description, lib.Functions[0].Flags[0])
+			err := client.FunctionLoad(ctx, libCode).Err()
+
+			Expect(err).NotTo(HaveOccurred())
+
+			r, err := client.FunctionStats(ctx).Result()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(r.Engines)).To(Equal(1))
+			Expect(r.Running()).To(BeFalse())
+
+			started := make(chan bool)
+			done := make(chan bool)
+			go func() {
+				defer GinkgoRecover()
+
+				started <- true
+				callResult := client.FCall(ctx, lib.Functions[0].Name, []string{})
+				Expect(callResult.Err()).NotTo(HaveOccurred())
+				done <- true
+			}()
+
+			<-started
+			time.Sleep(1 * time.Second)
+			r, err = client.FunctionStats(ctx).Result()
+
+			select {
+			case <-done:
+				Expect(err).NotTo(HaveOccurred())
+				Expect(len(r.Engines)).To(Equal(1))
+				rs, isRunning := r.RunningScript()
+				Expect(isRunning).To(BeTrue())
+				Expect(rs.Name).To(Equal(lib.Functions[0].Name))
+				Expect(rs.Duration > 0).To(BeTrue())
+			case <-time.After(10 * time.Second):
+				log.Println("Long-running function call timed out")
+			}
```

Comment:
I didn't understand the test here. Why do we need to wait for the "done" signal? The FCall command will block until LUA execution is complete, but when the Lua-time-limit is reached, Redis can accept the "function kill/stats" command. If FCall takes one minute to execute, this test will time out. Additionally, we should execute "function-kill" after the test ends because if LUA execution takes too long, it will continue to block subsequent tests.

Reply:
Good point about `function kill`. The function we're calling is relatively predictable and should never time out, but we should protect against it anyway! I'd have to change the function itself though, make it read-only, cause functions that have modified the keyspace in some way can't be killed. I'm on it!

Now, on to the more complicated issue :) I did some digging in the Redis source code and found that the`function stats` command lives in the same group as `function kill`, `quit`, `shutdown` commands. They all have the `CMD_ALLOW_BUSY` flag which lets them run while another command is running for a longer time. Otherwise, we would never get any results under the `running_script` field.

My point in using the `done` sync was to wait for fcall to complete before moving on to the next test, but your idea is better, we can just kill the function after we run the checks and that should force the goroutine to return.

Let me do these fixes and we can maybe look at the result again, see how can we make this pass.

Reply:
Btw, I saw that https://github.com/redis/go-redis/pull/2465 was closed in the meantime, so I went ahead and implemented `function kill` in this same PR, cause I needed it.